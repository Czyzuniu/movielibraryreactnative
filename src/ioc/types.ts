const InjectableTypes = {
  GetPopularMoviesUseCase: Symbol.for('GetPopularMoviesUseCase'),
  FindMovieByTextUseCase: Symbol.for('FindMovieByTextUseCase'),
  GetMovieByIdUseCase: Symbol.for('GetMovieByIdUseCase'),
  MovieRepository: Symbol.for('MovieRepository'),
  MovieRemoteDataSource: Symbol.for('MovieRemoteDataSource'),
  AxiosClient: Symbol.for('AxiosClient'),
  AsyncStorageRepository: Symbol.for('AsyncStorageRepository'),
};

export {InjectableTypes};
