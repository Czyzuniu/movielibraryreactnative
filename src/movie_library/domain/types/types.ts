export enum ProductCategoryType {
  Electronics = 'electronics',
  Jewelery = 'jewelery',
  ClothesWomen = "women's clothing",
  ClothesMen = "men's clothing",
}

export interface IUseCase<P, R> {
  execute: (params: P) => R;
}
