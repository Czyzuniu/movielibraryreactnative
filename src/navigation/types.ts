import {StackNavigationProp} from "@react-navigation/stack";
import {BottomTabNavigationProp} from "@react-navigation/bottom-tabs";

type PopularMoviesStackParamList = {
  PopularMovies: undefined,
  Common: StackNavigationProp<CommonStackParamList>
}

type CreatedListsStackStackParamList = {
  MyLists: undefined,
  CreateList: undefined,
  Common: StackNavigationProp<CommonStackParamList>
}

type SettingsStackTypeParamList = {
  Settings: undefined
}

type LoggedInBottomNavigationTabParamList = {
  PopularMoviesStack: StackNavigationProp<PopularMoviesStackParamList>;
  CreatedListsStack: StackNavigationProp<CreatedListsStackStackParamList>;
  SettingsStack: StackNavigationProp<SettingsStackTypeParamList>;
};

type LoginStackParamList = {
  Login: undefined
}

type RootStackParamList = {
  LoggedOut: StackNavigationProp<LoginStackParamList>
  LoggedIn: BottomTabNavigationProp<LoggedInBottomNavigationTabParamList>;
};

type CommonStackParamList = {
  ViewMovie: {movieId: string; title: string};
  ViewMovieHomePageWebView: {movieUrl: string};
};

export {
  LoggedInBottomNavigationTabParamList,
  RootStackParamList,
  CommonStackParamList,
  PopularMoviesStackParamList,
  SettingsStackTypeParamList,
  LoginStackParamList,
  CreatedListsStackStackParamList
};
