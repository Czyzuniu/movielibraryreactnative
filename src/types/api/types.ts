type AuthTokenResponse = {
  success: boolean,
  expires_at: string,
  request_token: string
}

type AuthNewSessionResponse = {
  success: boolean,
  session_id: string
}

type AddToFavouritesRequestBody = {
  accountId: string,
  movieId: number,
  type: 'movie' | 'tv'
}

type AccountDetailsResponse = {
  "id": number,
  "name": string,
  "include_adult": boolean,
  "username": string
}

type LoginRequest = {
  username: string,
  password: string
  request_token: string
}

export { AuthTokenResponse, LoginRequest, AuthNewSessionResponse, AccountDetailsResponse, AddToFavouritesRequestBody}
