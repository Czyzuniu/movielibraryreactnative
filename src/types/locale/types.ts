export type LocaleSettings = {
  language: string
}
