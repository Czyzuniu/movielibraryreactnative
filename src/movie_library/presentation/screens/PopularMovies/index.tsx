import React, {useEffect, useState} from 'react';

import {Box, Button, Center} from 'native-base';
import {StackScreenProps} from '@react-navigation/stack';
import {HomeStackParamList} from '../../../../navigation/types';
import SearchBar from '../../../../base/presentation/components/SearchBar';
import LanguageModal from '../../../../base/presentation/components/LanguageModal';
import {useGetPopularMoviesQuery} from "../../../../redux/services/movies";
import MovieDto from "../../../infrastructure/models/MovieDto";
import MovieVerticalCardList from "../../components/MovieVerticalCardList";
import Icon from "react-native-vector-icons/FontAwesome";

type Props = StackScreenProps<HomeStackParamList, 'Home'>;

export default function PopularMovies({navigation}: Props) {
  const [search, setSearch] = useState('');
  const [page, setPage] = useState(1);
  const {
    data = {
      results: []
    }, isFetching
  } = useGetPopularMoviesQuery(page)
  const [items, setItems] = useState<Array<MovieDto>>(data?.results || []);

  useEffect(() => {
    if (data.results.length) {
      const hasDuplicate = items.find(movie => movie.id === data.results[0].id);
      if (hasDuplicate) {
        setItems(data.results)
      } else {
        setItems(items.concat(data.results))
      }
    }
  }, [data])

  return (
    <Box
      _light={{
        backgroundColor: 'white.300',
      }}
      _dark={{
        backgroundColor: 'muted.800',
      }}
      flex={1}
    >
      <Center flex={0.1} m={1.5}>
        <SearchBar onSearch={setSearch}/>
      </Center>
      <Box flex={0.9}>
        <MovieVerticalCardList data={items} ListFooterComponent={() => {
          if (items.length) {
            return (
              <Center flex={0.2} m={5}>
                <Button
                  variant={'solid'}
                  isLoading={isFetching}
                  spinnerPlacement="end"
                  onPress={() => {
                    setPage(page + 1)
                  }}
                  leftIcon={<Icon name={'chevron-circle-down'} color={'white'}/>} w={250}>Load more...
                </Button>
              </Center>
            )
          }
          return null
        }}/>
      </Box>
      <LanguageModal/>
    </Box>
  );
}
