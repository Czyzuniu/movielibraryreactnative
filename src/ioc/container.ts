import 'reflect-metadata';
import {Container, ContainerModule} from 'inversify';
import AxiosClient from '../base/infrastructure/AxiosClient';
import {InjectableTypes} from './types';
import {
  MoviesRemoteApiDataSource,
  MoviesRemoteApiDataSourceImpl,
} from '../movie_library/infrastructure/datasource/MoviesRemoteApiDataSource';
import {MovieRepository} from '../movie_library/domain/repository/MovieRepository';
import MovieRepositoryImpl from '../movie_library/infrastructure/repository/MovieRepositoryImpl';
import GetPopularMoviesUseCase from '../movie_library/domain/usecase/GetPopularMoviesUseCase';
import GetMovieByIdUseCase from '../movie_library/domain/usecase/GetMovieByIdUseCase';
import {AsyncStorageRepository} from '../async_storage/domain/repository/AsyncStorageRepository';
import AsyncStorageRepositoryImpl from '../async_storage/infrastructure/AsyncStorageRepositoryImpl';

const myContainer = new Container();

const useCases = new ContainerModule(bind => {
  bind<GetPopularMoviesUseCase>(InjectableTypes.GetPopularMoviesUseCase).to(
    GetPopularMoviesUseCase,
  );
  bind<GetMovieByIdUseCase>(InjectableTypes.GetMovieByIdUseCase).to(
    GetMovieByIdUseCase,
  );
});

const repositories = new ContainerModule(bind => {
  bind<MovieRepository>(InjectableTypes.MovieRepository)
    .to(MovieRepositoryImpl)
    .inSingletonScope();

  bind<AsyncStorageRepository>(InjectableTypes.AsyncStorageRepository)
    .to(AsyncStorageRepositoryImpl)
    .inSingletonScope();
});

const datasources = new ContainerModule(bind => {
  bind<MoviesRemoteApiDataSource>(InjectableTypes.MovieRemoteDataSource)
    .to(MoviesRemoteApiDataSourceImpl)
    .inSingletonScope();
});

const clients = new ContainerModule(bind => {
  bind<AxiosClient>(InjectableTypes.AxiosClient)
    .to(AxiosClient)
    .inSingletonScope();
});

myContainer.load(useCases, repositories, datasources, clients);

export {myContainer};
