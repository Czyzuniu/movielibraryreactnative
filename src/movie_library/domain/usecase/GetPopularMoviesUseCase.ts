import {MovieRepository} from '../repository/MovieRepository';
import SimplisticMovie from '../entity/SimplisticMovie';
import {InjectableTypes} from '../../../ioc/types';
import {inject, injectable} from 'inversify';

@injectable()
export default class GetPopularMoviesUseCase {
  private repository: MovieRepository;

  constructor(
    @inject(InjectableTypes.MovieRepository) repository: MovieRepository,
  ) {
    this.repository = repository;
  }

  async execute(
    page: number,
    searchQuery: string,
  ): Promise<Array<SimplisticMovie>> {
    return this.repository.getPopularMovies(page, searchQuery);
  }
}
