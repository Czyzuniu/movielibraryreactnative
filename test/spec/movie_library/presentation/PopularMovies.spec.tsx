import {fireEvent, render, waitFor} from '@testing-library/react-native';
import React from 'react';
import ContextWrapper from '../../../utils/ContextWrapper';
import PopularMovies from '../../../../src/movie_library/presentation/screens/PopularMovies';
import {mockNavigation} from '../../../utils/mockNavigation';
import GetPopularMoviesUseCase from '../../../../src/movie_library/domain/usecase/GetPopularMoviesUseCase';
import {mockDeep} from 'jest-mock-extended';
import {when} from 'jest-when';
import {myContainer} from '../../../../src/ioc/container';
import {InjectableTypes} from '../../../../src/ioc/types';
import TestObjects from '../../../utils/TestObjects';
import {Alert} from 'react-native';
import {act} from 'react-test-renderer';

describe('PopularMovies Screen Tests', () => {
  let popularMoviesUseCaseMock: GetPopularMoviesUseCase;

  beforeEach(() => {
    popularMoviesUseCaseMock = mockDeep<GetPopularMoviesUseCase>();
    jest.spyOn(myContainer, 'get').mockImplementation(() => jest.fn());
    when(myContainer.get)
      .calledWith(InjectableTypes.GetPopularMoviesUseCase)
      .mockReturnValue(popularMoviesUseCaseMock);
  });

  it('should render and match snapshot', () => {
    const {toJSON} = render(
      <ContextWrapper>
        <PopularMovies {...mockNavigation()} />
      </ContextWrapper>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should fetch call the usecase to fetch the popular movies', async () => {
    const movies = [
      TestObjects.SimplisticMovie(),
      TestObjects.SimplisticMovie(),
    ];

    when(popularMoviesUseCaseMock.execute).mockResolvedValue(movies);

    const {getByTestId} = render(
      <ContextWrapper>
        <PopularMovies {...mockNavigation()} />
      </ContextWrapper>,
    );

    await waitFor(() =>
      expect(getByTestId('POPULAR_MOVIES_FLAT_LIST').props.data).toStrictEqual(
        movies,
      ),
    );
  });

  it('should navigate to ViewMovie if a movie card is tapped', async () => {
    const movie = TestObjects.SimplisticMovie();
    const movies = [movie];
    const navigation = mockNavigation();
    when(popularMoviesUseCaseMock.execute).mockResolvedValue(movies);

    const {getByTestId} = render(
      <ContextWrapper>
        <PopularMovies {...navigation} />
      </ContextWrapper>,
    );

    await waitFor(() =>
      expect(getByTestId('POPULAR_MOVIES_FLAT_LIST').props.data).toStrictEqual(
        movies,
      ),
    );

    fireEvent.press(getByTestId('MOVIE_CARD_' + movie.id));

    expect(navigation.navigation.navigate).toBeCalledWith('ViewMovie', {
      movieId: movie.id,
      title: movie.title,
    });
  });

  it('should display alert if fails to fetch movies', async () => {
    jest.spyOn(Alert, 'alert');

    const exception = new Error('error');
    const navigation = mockNavigation();
    when(popularMoviesUseCaseMock.execute).mockRejectedValue(exception);

    render(
      <ContextWrapper>
        <PopularMovies {...navigation} />
      </ContextWrapper>,
    );

    await waitFor(() => expect(Alert.alert).toBeCalledWith(exception.message));
  });

  it('should refetch when reached the end', async () => {
    const movie = TestObjects.SimplisticMovie();
    const movies = [movie];
    const navigation = mockNavigation();
    when(popularMoviesUseCaseMock.execute).mockResolvedValue(movies);

    const {getByTestId} = render(
      <ContextWrapper>
        <PopularMovies {...navigation} />
      </ContextWrapper>,
    );

    const flatlist = getByTestId('POPULAR_MOVIES_FLAT_LIST');

    await waitFor(() => expect(flatlist.props.data).toStrictEqual(movies));

    await act(async () => {
      flatlist.props.onEndReached();
    });

    await waitFor(() =>
      expect(popularMoviesUseCaseMock.execute).toBeCalledWith(2, ''),
    );
  });
});
