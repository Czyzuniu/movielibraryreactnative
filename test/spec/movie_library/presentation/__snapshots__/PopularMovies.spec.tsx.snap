// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopularMovies Screen Tests should render and match snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    dataSet={Object {}}
    style={
      Object {
        "backgroundColor": "#262626",
        "flex": 1,
      }
    }
  >
    <View
      dataSet={Object {}}
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flex": 0.1,
          "justifyContent": "center",
          "marginBottom": 6,
          "marginLeft": 6,
          "marginRight": 6,
          "marginTop": 6,
        }
      }
    >
      <View
        dataSet={Object {}}
        isFocused={false}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#262626",
            "borderColor": "#262626",
            "borderRadius": 4,
            "borderWidth": 1,
            "flex": 1,
            "flexDirection": "row",
            "marginLeft": 20,
            "marginRight": 20,
            "overflow": "hidden",
            "width": "75%",
          }
        }
      >
        <TextInput
          accessible={true}
          dataSet={Object {}}
          editable={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          placeholder="Search"
          placeholderTextColor="#525252"
          secureTextEntry={false}
          selectionColor="#fafaf9"
          style={
            Object {
              "backgroundColor": "transparent",
              "color": "#fafafa",
              "flex": 1,
              "fontFamily": undefined,
              "fontSize": 14,
              "fontStyle": "normal",
              "fontWeight": "400",
              "height": "100%",
              "paddingBottom": 8,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 8,
              "width": "100%",
            }
          }
          value=""
        />
      </View>
    </View>
    <View
      dataSet={Object {}}
      style={
        Object {
          "flex": 0.9,
        }
      }
    >
      <RCTScrollView
        ListFooterComponent={[Function]}
        contentContainerStyle={
          Object {
            "alignSelf": "center",
          }
        }
        data={Array []}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onEndReachedThreshold={0.1}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        refreshControl={
          <RefreshControlMock
            onRefresh={[Function]}
            refreshing={true}
          />
        }
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        testID="POPULAR_MOVIES_FLAT_LIST"
        viewabilityConfigCallbackPairs={Array []}
      >
        <RCTRefreshControl />
        <View>
          <View
            onLayout={[Function]}
          />
        </View>
      </RCTScrollView>
    </View>
  </View>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "display": "none",
        },
      ]
    }
  >
    <View
      collapsable={false}
      entryDuration={200}
      exitDuration={150}
      pointerEvents="none"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
            Object {
              "translateX": 0,
            },
            Object {
              "scale": 1,
            },
            Object {
              "scaleX": 1,
            },
            Object {
              "scaleY": 1,
            },
            Object {
              "rotate": "0deg",
            },
          ],
        }
      }
    >
      <View
        accessible={false}
        collapsable={false}
        dataSet={Object {}}
        focusable={true}
        importantForAccessibility="no"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(0, 0, 0)",
            "bottom": 0,
            "left": 0,
            "opacity": 0.3,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      />
    </View>
    <View
      collapsable={false}
      entryDuration={200}
      exitDuration={100}
      pointerEvents="none"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
          "transform": Array [
            Object {
              "translateY": 0,
            },
            Object {
              "translateX": 0,
            },
            Object {
              "scale": 1,
            },
            Object {
              "scaleX": 1,
            },
            Object {
              "scaleY": 1,
            },
            Object {
              "rotate": "0deg",
            },
          ],
        }
      }
    >
      <View
        closeOnOverlayClick={true}
        dataSet={Object {}}
        pointerEvents="box-none"
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "height": "100%",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <View
          accessibilityViewIsModal={true}
          aria-modal="true"
          dataSet={Object {}}
          onAccessibilityEscape={[Function]}
          style={
            Object {
              "backgroundColor": "#262626",
              "borderRadius": 8,
              "elevation": 2,
              "flex": 1,
              "maxHeight": 1184,
              "maxWidth": 600,
              "overflow": "hidden",
              "shadowColor": "#000000",
              "shadowOffset": Object {
                "height": 1,
                "width": 0,
              },
              "shadowOpacity": 0.2,
              "shadowRadius": 1.41,
              "width": "75%",
            }
          }
        >
          <View
            dataSet={Object {}}
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <View
              style={
                Object {
                  "height": 50,
                  "width": "100%",
                }
              }
            >
              <RNCPicker
                items={
                  Array [
                    Object {
                      "label": "en-GB",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": "en-GB",
                    },
                    Object {
                      "label": "fr-FR",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": "fr-FR",
                    },
                    Object {
                      "label": "pl",
                      "testID": undefined,
                      "textColor": undefined,
                      "value": "pl",
                    },
                  ]
                }
                numberOfLines={1}
                onChange={[Function]}
                selectedIndex={0}
                style={
                  Array [
                    Object {
                      "height": 216,
                    },
                    undefined,
                  ]
                }
              />
            </View>
          </View>
          <View
            accessibilityRole="button"
            accessible={true}
            collapsable={false}
            colorScheme="success"
            dataSet={Object {}}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#16a34a",
                "borderRadius": 4,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginBottom": 8,
                "marginLeft": 8,
                "marginRight": 8,
                "marginTop": 8,
                "paddingBottom": 10,
                "paddingLeft": 12,
                "paddingRight": 12,
                "paddingTop": 10,
              }
            }
          >
            <View
              dataSet={Object {}}
              style={
                Object {
                  "alignItems": "center",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
              test={true}
            >
              <View
                dataSet={Object {}}
                style={
                  Object {
                    "flex": 1,
                  }
                }
              >
                <Text
                  dataSet={Object {}}
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "color": "#fafafa",
                      "fontFamily": undefined,
                      "fontSize": 14,
                      "fontStyle": "normal",
                      "fontWeight": "400",
                      "letterSpacing": 0,
                      "lineHeight": 21,
                      "textDecorationLine": undefined,
                    }
                  }
                >
                  common:Save
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
